src/Loggers/CMakeLists.txt

```diff
@@ -2,9 +2,9 @@ add_headers_and_sources(loggers .)
 
 # Standard version depends on DBMS and works with text log
 add_library(loggers ${loggers_sources} ${loggers_headers})
-target_compile_definitions(loggers PUBLIC WITH_TEXT_LOG=1)
 target_link_libraries(loggers PRIVATE dbms clickhouse_common_io)
 
 # Lightweight version doesn't work with textlog and also doesn't depend on DBMS
 add_library(loggers_no_text_log ${loggers_sources} ${loggers_headers})
+target_compile_definitions(loggers_no_text_log PUBLIC WITHOUT_TEXT_LOG=1)
```

Comment:
@rschu1ze AFAIU this line is supposed to fix the build, but to be honest I don't understand why. For me it seems to have absolutely the same behaviour as before because `#ifdef WITH_TEXT_LOG` was replaced with `ifndef WITHOUT_TEXT_LOG`.

Reply:
i figure that the main binary include 'Loggers.h' in different way, in most case, the include path is implicated by loggers library. But some directly include the 'Loggers.h'. so the `WITH_TEXT_LOG` definition differ in the main binary.
I just reverse `WITH_TEXT_LOG` to `WITHOUT_TEXT_LOG`. so in the main library all 'Loggers.h'  has no `WITHOUT_TEXT_LOG`. it works out.

Reply:
The fix is a bit weird but generally okay, IMHO.

It would be great to understand the problem in more detail.

@Lucky-Chang Could you please give the detailed steps for reproduction: What is the exact command line for cmake? Please try to use homebrew's LLVM/Clang/CMake as described in the [documentation](https://clickhouse.com/docs/en/development/build-osx) instead of Apple's XCode LLVM/Clang (as used in #40398).

> i figure that the main binary include 'Loggers.h' in different way, in most case, the include path is implicated by loggers library. But some directly include the 'Loggers.h'. so the WITH_TEXT_LOG definition differ in the main binary.

- By "main binary", do you mean programs/server/Server.cpp?
- What do you mean "the loggers library implicates the include path"? More specifically, which files get included exactly and where exactly does that happen?
- Could you give an example for the files which "directly include Loggers.h"?

I think I have a clue about the cause of the problem. Sorry for asking all these questions but depending on the details, there could be potentially a better (more consistent) fix.

Reply:
It took me quite some time to understand how it fix the code. The simple answer `Loggers` class definition violates ODR (introduced in #35031).
Explanaition:
1. `Loggers` class is build in 2 binaries: `loggers` and `loggers_no_text_log`.
2. `loggers_no_text_log` is used only for clickhouse-keeper.
3. Everything else is linked against `loggers`, but compiled without `WITH_TEXT_LOG` macro.

For example, in BaseDeamon.h we have `#include <Loggers/Loggers.h>`. BaseDeamon.cpp is never compiled with `WITH_TEXT_LOG` macro defined, but linked with `loggers`.